### PR TITLE
Improve function handling

### LINKED ISSUE #231: API docs need to handle functions better
E.g. https://docs.quantum-computing.ibm.com/api/qiskit/compiler

Screenshot 2023-10-23 at 6 02 53 PM

The header should better show what the function name is. And the right sidebar should be more useful:
 
Screenshot 2023-10-23 at 6 03 16 PM

But it is important the function still shows the whole import path because we have some modules right now that have functions that have functions with submodules, e.g. the module page for `my_module` has a function `my_module.submodule.my_func` -- you have to show the whole import path for the docs to be correct.

### PR DESCRIPTION
**Pull request by frankharkins**
This PR improves how the API generation script handles functions.

### Demote some `h2` tags to `strong`

We currently convert all `.rubric` elements to `h2`, but this makes some headings higher-level than they should be (specifically "Example" and "References" headings). For an example see [api/qiskit/converters](https://docs.quantum.ibm.com/api/qiskit/converters).

Screenshot 2024-01-23 at 17 37 33

This PR converts rubrics to `strong` (rather than `h2`) if the name is "Example(s)" or "Reference(s)". This is a bit hacky, but it seems to work and I can't think of a better way of doing it. The Sphinx LaTeX writer also takes a [similar approach](https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-rubric).

I've regenerated the docs using this new logic and I can see no false positives (that is, cases in which we use `strong` where we should use `h2`).

### Add function names as headings

Functions currently have no headings. This PR adds a `h3` with the function name, which improves readability, gives anchor tag icons, and will add more functions to the right sidebar.

I've also included some logic to avoid adding function names when the page is already dedicated to the function (such as [least_busy](https://docs.quantum.ibm.com/api/qiskit-ibm-provider/qiskit_ibm_provider.least_busy)).

One quirk with this I spotted was some function names not appearing in the sidebar in `/api/qiskit/utils` (see `add_deprecation_to_docstring` in the screenshot). This is possibly because the title hierarchy skips `h2`.

Screenshot 2024-01-23 at 18 00 37

***

Closes #231

### CODE REVIEW

File: scripts/lib/api/processHtml.ts [hunk 1]
```diff
@@ -194,12 +194,21 @@ export function convertRubricsToHeaders(
...
-  //  if always using <h2> makes sense.
+  // A rubric is "a paragraph heading that is not used to create a table
+  // of contents node". Depending on the heading, this should be either <h2> or
+  // <strong>
+  function appropriateHtmlTag(html: string | null) {
+    html = String(html);
+    if (html.match(/^Examples?$/) || html.match(/^References$/)) {
+      return "strong";
+    }
+    return "h2";
```
Comment (Eric-Arellano) [inline scripts/lib/api/processHtml.ts:205]:
```suggestion
    return html.match(/^Examples?$/) || html.match(/^References$/) ? "strong" : "h2";
```

File: scripts/lib/api/processHtml.ts [hunk 2]
```diff
@@ -194,12 +194,21 @@ export function convertRubricsToHeaders(
   $: CheerioAPI,
   $main: Cheerio<any>,
 ): void {
```
Comment (Eric-Arellano) [inline scripts/lib/api/processHtml.ts:196]: This function would be useful to add unit tests. I only had added them before due to time.

File: scripts/lib/api/processHtml.ts [hunk 3]
```diff
@@ -353,7 +362,15 @@ export function processMembersAndSetMeta(
 
         if (apiType === "function") {
           findByText($, $main, "em.property", "function").remove();
-          return `<span class="target" id="${id}"/><p><code>${$child.html()}</code>${github}</p>`;
+          const funcDescription = `<span class="target" id="${id}"/><p><code>${$child.html()}</code>${github}</p>`;
+
+          const pageHeading = $dl.siblings("h1").text();
+          const funcName = id.split(".").slice(-1)[0];
+          if (id.endsWith(pageHeading) && pageHeading != "") {
+            // Page is already dedicated to function; no heading needed
```
Comment (Eric-Arellano) [inline scripts/lib/api/processHtml.ts:370]: Nice comment!

File: scripts/lib/api/processHtml.ts [hunk 4]
```diff
@@ -353,7 +362,15 @@ export function processMembersAndSetMeta(
 
         if (apiType === "function") {
```
Comment (Eric-Arellano) [inline scripts/lib/api/processHtml.ts:363]: This would also be great to test. I know we don't have any tests for this function overall and that it's a bit tricky to test - I hadn't gotten to it because it was more complex hah. 

1. Totally okay to only test `function` and not worry about the others
2. If the overall function is too hard to test, it might be easier to extract a helper function and test that.

**Review by Eric-Arellano — COMMENTED** [review]
This looks great - excellent improvement! Please do not merge until we've finished regenerating all historical API docs with prior queued changes.

File: scripts/lib/api/processHtml.ts (hunk 4 above)
Comment (frankharkins) [reply scripts/lib/api/processHtml.ts:362]: Good call, I've only added tests for `function` in this PR to keep it focused, but I can follow up with a PR to add others.

File: scripts/lib/api/processHtml.test.ts [hunk 5]
```diff
@@ -247,6 +249,26 @@ test("replaceSourceLinksWithGitHub()", () => {
...
+    <p class="rubric">Reference</p>
+    <p class="rubric">Attributes</p>
+    <p class="rubric">Methods</p>
+    <p class="rubric">Simple examples</p>
+    <p class="rubric">Example code</p>`);
+  convertRubricsToHeaders(doc.$, doc.$main);
+  doc.expectHtml(`<strong>Example</strong>
+    <strong>Examples</strong>
+    <strong>References</strong>
+    <h2>Reference</h2>
```
Comment (Eric-Arellano) [inline scripts/lib/api/processHtml.test.ts:265]: This should be `<strong>` because it's "member-specific", similar to an example. Rather than signifying the start of a new section with its own "code members", this is only added info about the current member.

File: scripts/lib/api/processHtml.test.ts [hunk 6]
```diff
@@ -247,6 +249,26 @@ test("replaceSourceLinksWithGitHub()", () => {
...
+    <h2>Attributes</h2>
+    <h2>Methods</h2>
+    <h2>Simple examples</h2>
+    <h2>Example code</h2>`);
```
Comment (Eric-Arellano) [inline scripts/lib/api/processHtml.test.ts:269]: These should be `<strong>` for the same reason as Reference.

File: scripts/lib/api/processHtml.test.ts [hunk 7]
```diff
@@ -247,6 +249,26 @@ test("replaceSourceLinksWithGitHub()", () => {
   );
 });
 
+test("convertRubricsToHeaders()", () => {
```
Comment (Eric-Arellano) [inline scripts/lib/api/processHtml.test.ts:252]: Excellent test. This is so helpful! Thank you for finding all the known edge cases.

File: scripts/lib/api/processHtml.test.ts [hunk 8]
```diff
@@ -313,3 +335,88 @@ describe("prepareGitHubLink()", () => {
...
+<dd><p>Build a <a class="reference internal" href="../stubs/qiskit.dagcircuit.DAGCircuit.html#qiskit.dagcircuit.DAGCircuit" title="qiskit.dagcircuit.DAGCircuit"><code class="xref py py-class docutils literal notranslate"><span class="pre">DAGCircuit</span></code></a> object from a <a class="reference internal" href="../stubs/qiskit.circuit.QuantumCircuit.html#qiskit.circuit.QuantumCircuit" title="qiskit.circuit.QuantumCircuit"><code class="xref py py-class docutils literal notranslate"><span class="pre">QuantumCircuit</span></code></a>.</p>
+<dl class="field-list simple">
+<dt class="field-odd">Parameters<span class="colon">:</span></dt>
+<dd class="field-odd"><ul class="simple">
+<li><p><strong>circuit</strong> – the input circuit.</p></li>
+<li><p><strong>copy_operations</strong> – Deep copy the operation objects in the QuantumCircuit for the output DAGCircuit.</p></li>
+</ul>
+</dd>`;
+    const doc = Doc.load(html);
+    processMembersAndSetMeta(doc.$, doc.$main, {});
```
Comment (Eric-Arellano) [inline scripts/lib/api/processHtml.test.ts:354]: It'd be good to test that the `meta` is set properly. You'd need to store it as a dedicated variable so that you can then `expect()` what it looks like after.

**Review by Eric-Arellano — COMMENTED** [review]
💜 This will be ready to merge once Arnau puts up the last PR for Qiskit 0.19.

File: scripts/lib/api/processHtml.test.ts (hunk 6 above)
Comment (frankharkins) [reply scripts/lib/api/processHtml.test.ts:269]: What should we match then? Anything containing example? Or should we go the other route and have everything be `strong`, with `Methods` and `Attributes` being the exceptions?

Comment (frankharkins) [reply scripts/lib/api/processHtml.test.ts:269]: > should we [...] have everything be `strong`, with `Methods` and `Attributes` being the exceptions?

With another look at the docs, I think that might be best. All the other rubrics I can see should be `strong` IMO.

<details>
 <summary>Other rubrics</summary>

```html
python/sources/qiskit/0.45.2/artifact/apidoc/pulse.html
2120:<p class="rubric">Notes</p>
2178:<p class="rubric">Notes</p>

python/sources/qiskit/0.45.2/artifact/stubs/qiskit.pulse.ScheduleBlock.html
53:<p class="rubric">Pulse Builder</p>
57:<p class="rubric">Alignment contexts</p>
68:<p class="rubric">Nested blocks</p>
75:<p class="rubric">Unsupported operations</p>
83:<p class="rubric">Subroutine</p>
95:<p class="rubric">Program Scoping</p>

python/sources/qiskit/0.45.2/artifact/stubs/qiskit.pulse.Schedule.html
116:<p class="rubric">Notes</p>

python/sources/qiskit/0.45.2/artifact/stubs/qiskit.transpiler.passes.ConsolidateBlocks.html
51:<p class="rubric">Notes</p>

python/sources/qiskit/0.45.2/artifact/stubs/qiskit.pulse.library.SymbolicPulse.html
66:<p class="rubric" id="symbolic-pulse-envelope">Envelope function</p>
82:<p class="rubric" id="symbolic-pulse-constraints">Constraint functions</p>
92:<p class="rubric" id="symbolic-pulse-eval-condition">Maximum amplitude validation</p>
169:<p class="rubric" id="symbolic-pulse-serialize">Serialization</p>

python/sources/qiskit/0.45.2/artifact/stubs/qiskit.result.QuasiDistribution.html
184:<p class="rubric">Notes</p>

python/sources/qiskit/0.45.2/artifact/stubs/qiskit.quantum_info.SparsePauliOp.html
58:<p class="rubric">Data type of coefficients</p>

python/sources/qiskit/0.45.2/artifact/stubs/qiskit.circuit.QuantumCircuit.html
1291:<p class="rubric">Notes</p>
1522:<p class="rubric">Notes</p>
2516:<p class="rubric">Notes</p>

python/sources/qiskit/0.45.2/artifact/stubs/qiskit.transpiler.passes.StochasticSwap.html
49:<p class="rubric">Notes</p>

python/sources/qiskit/0.45.2/artifact/stubs/qiskit.transpiler.passes.ConstrainedReschedule.html
69:<p class="rubric">Notes</p>

python/sources/qiskit/0.45.2/artifact/stubs/qiskit.algorithms.optimizers.NFT.html
64:<p class="rubric">Notes</p>

python/sources/qiskit/0.45.2/artifact/stubs/qiskit.transpiler.passes.AlignMeasures.html
91:<p class="rubric">Notes</p>
```
</details>

Comment (Eric-Arellano) [reply scripts/lib/api/processHtml.test.ts:269]: 100% agreed.

**Comment by arnaucasau** [comment]
All Qiskit versions have been regenerated so this can be merged when you both think it's ready :+1:. Thank you so much @frankharkins for waiting!

**Comment by Eric-Arellano** [comment]
You'll need to regenerate the Qiskit 0.45 docs. You can run this:

```
npm run gen-api -- -p qiskit -v 0.45.2 -a https://github.com/Qiskit/qiskit/actions/runs/7626116665/artifacts/1188676326
```

File: scripts/lib/api/processHtml.test.ts (hunk 8 above)
Comment (frankharkins) [reply scripts/lib/api/processHtml.test.ts]: Thanks! Done in 52110cea467336c146c74027b2a111412253d4c9

File: scripts/lib/api/processHtml.test.ts (hunk 6 above)
Comment (frankharkins) [reply scripts/lib/api/processHtml.test.ts]: Done in 419645b5d7356c00d53fb26059186b98b2656690

**Review by Eric-Arellano — APPROVED** [review]
🚀 Arnau is working on automating regeneration of historical API docs. So he can help with applying this fix to the historical docs.

Great job with this! Huge improvement.
